drivers/spi/spi-rp2040-gpio-bridge.c

```diff
@@ -0,0 +1,1219 @@
+	rp2040_gbdg_gpio_dir_in(gc, offset);
+	rp2040_gbdg_gpio_set(gc, offset, 0);
+
+	if (!priv_data->gpio_requested) {
+		ret = pm_runtime_put_autosuspend(&priv_data->client->dev);
+		if (ret) {
+			dev_err(&priv_data->client->dev,
+				"%s(%u) unable to put_autosuspend\n", __func__,
+				offset);
+		}
+	}
+}
+
+static int rp2040_gbdg_gpio_get_direction(struct gpio_chip *gc,
+					  unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	return (priv_data->gpio_direction & (1 << (offset + 8))) ?
+		       GPIO_LINE_DIRECTION_IN :
+		       GPIO_LINE_DIRECTION_OUT;
+}
+
+static int rp2040_gbdg_gpio_dir_in(struct gpio_chip *gc, unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	priv_data->gpio_direction |= (1 << (offset + 8));
+
+	return rp2040_gbdg_10byte_cmd(client, CMD_GPIO_OE,
+				      ~priv_data->gpio_direction,
+				      priv_data->gpio_direction);
+}
+
+static int rp2040_gbdg_gpio_dir_out(struct gpio_chip *gc, unsigned int offset,
+				    int value)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	u32 pattern;
+	int ret;
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	pattern = (1 << (offset + 8));
+
+	ret = rp2040_gbdg_10byte_cmd(client, CMD_GPIO_ST_CL,
+				     value ? pattern : 0, !value ? pattern : 0);
+	if (ret) {
+		dev_err(&client->dev, "%s(%u, %d) could not ST_CL\n", __func__,
+			offset, value);
+		return ret;
+	}
+
+	priv_data->gpio_direction &= ~pattern;
+	ret = rp2040_gbdg_10byte_cmd(client, CMD_GPIO_OE,
+				     ~priv_data->gpio_direction,
+				     priv_data->gpio_direction);
+
+	return ret;
+}
+
+static int rp2040_gbdg_gpio_get(struct gpio_chip *gc, unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	struct rp2040_gbdg_device_info info;
+	int ret;
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	ret = rp2040_gbdg_get_device_info(client, &info);
+	if (ret)
+		return ret;
+
+	return info.status & (1 << (offset + 8)) ? 1 : 0;
+}
+
+static void rp2040_gbdg_gpio_set(struct gpio_chip *gc, unsigned int offset,
+				 int value)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	u32 pattern;
+
+	if (offset >= NUM_GPIO)
+		return;
+
+	pattern = (1 << (offset + 8));
+	rp2040_gbdg_10byte_cmd(client, CMD_GPIO_ST_CL, value ? pattern : 0,
+			       !value ? pattern : 0);
+}
+
+static int rp2040_gbdg_get_regulator(struct device *dev,
+				     struct rp2040_gbdg *rp2040_gbdg)
+{
+	struct regulator *reg = devm_regulator_get(dev, "power");
+
+	if (IS_ERR(reg))
+		return PTR_ERR(reg);
+
+	rp2040_gbdg->regulator = reg;
+
+	return 0;
+}
+
+static void rp2040_gbdg_parse_dt(struct rp2040_gbdg *rp2040_gbdg)
```

Comment:
Is it reasonable that this has no return value? It has many points where it returns early after an error.

Reply:
Only the parsing of the 'bypass-cache' property in rp2040_gbdg_parse_dt() is required for all platforms. This cannot fail due to the API of of_property_read_bool() returning the bool.

Any failure in the rest of rp2040_gbdg_parse_dt() is not an issue as the driver will use I2C to transfer blocks of data as a fallback (rather than fast-xfer using RP1 bit-bashing). In fact, we expect failure on platforms without RP1.

If we were to return a value from rp2040_gbdg_parse_dt, in it's current form, it would only be informational and not to be acted upon (other than perhaps a dev_info).

I'm happy to refactor rp2040_gbdg_parse_dt() if we can think of a cleaner/easier to understand API. Alternatively, I could comment the function as it's perhaps not sufficiently clear in its current form.

Reply:
It's OK - if you're happy justifying it being as it is now, I'm happy to merge it